### PR TITLE
Return appid with an update-status noupdate if component is up to date

### PR DESCRIPTION
**Pull request by jumde**
Fix https://github.com/brave/brave-browser/issues/1319

## Description

Component updater expects `appid` and update-status `noupdate` if the component is up to date. If `appid` and `update-status` are not present in the response from the go-updater,  `UpdateCheckResult` for the component is set to `UPDATE_RESPONSE_NOT_FOUND` showing an error on the `brave://components` screen.
(https://cs.chromium.org/chromium/src/components/update_client/update_engine.cc?l=262)

### Before the fix:

screen shot 2019-01-23 at 5 53 28 pm

### After the fix:
screen shot 2019-01-23 at 5 54 10 pm

Please feel free to reach out if you have any questions.

auditors: @bsclifton, @bbondy

## Test Plan:

1. Component Install:

- Open Brave with a clean new profile
- [x] Navigate to `brave://components` and verify if all the components are installed correctly.

2. Component Updates:

- [x] Relaunch brave and verify that no component is showing a `Update Error`

3. Individual Updates:

- [x] Click on `Check for update` on individual components and verify no error is shown
    - [x] Brave Component
    - [x] Chromium Component

4. New Extensions:

- [x] Install any extension and verify it's working correctly

5. Update Extensions (update available) - Wait for Adblock to update

- [x] Wait for a component update and verify that the component is updated successfully.

6. Update Error
- [x] Turned off the go-update server, and verified that the components were not updating with the correct status `Update Error` 

## Known failures

1. https://github.com/brave/brave-browser/issues/3093 - DCHECK failure on extension install (repros on master)

### CODE REVIEW

**Comment by jumde** [comment]
Pending items:

- [ ] Update the `Test Plan` for component failure cases. When `go-updater` is not running.
- [ ] Check for any edge cases on other platforms (Linux, Windows)
- [ ] Verify the `Component Update` path works correctly (Adblock update)
- [ ] Verify `Old Web Store` extensions are updated correctly.

File: server/server_test.go [hunk 1]
```diff
@@ -170,7 +170,20 @@ func TestUpdateExtensions(t *testing.T) {
 
 	// Single extension same version
 	requestBody = lightThemeExtension("1.0.0")
-	expectedResponse = "<response protocol=\"3.1\" server=\"prod\"></response>"
+	expectedResponse = `<response protocol="3.1" server="prod">
+    <app appid="ldimlcelhnjgpjjemdjokpgeeikdinbm">
```
Comment (bbondy) [inline server/server_test.go]: Do you know, does the Google component update server return the same when there is no update?

File: extension/extension.go [hunk 2]
```diff
@@ -76,7 +77,11 @@ func (updateRequest *UpdateRequest) FilterForUpdates(allExtensionsMap *map[strin
 	for _, extensionBeingChecked := range *updateRequest {
 		foundExtension, ok := (*allExtensionsMap)[extensionBeingChecked.ID]
 		if ok {
-			if !foundExtension.Blacklisted && CompareVersions(extensionBeingChecked.Version, foundExtension.Version) < 0 {
+			status := CompareVersions(extensionBeingChecked.Version, foundExtension.Version)
+			if !foundExtension.Blacklisted && status <= 0 {
+				if status == 0 {
```
Comment (bbondy) [inline extension/extension.go]: Should this maybe also be done when status > 0?
What does Chrome do in the case you have a newer extension?

File: server/server_test.go (hunk 1 above)
Comment (jumde) [reply server/server_test.go]: Yes the response is similar. :
```
<response protocol=\"3.1\" server=\"prod\"><daystart elapsed_days=\"4404\" elapsed_seconds=\"46144\"/><app appid=\"giekcmmlnklenlaomppkphknjmnnpneh\" cohort=\"1:j5l:\" cohorthint=\"Auto\" cohortname=\"Auto\" status=\"ok\"><updatecheck status=\"noupdate\"/><ping status=\"ok\"/></app></response>
```

The response from `go-updater` does have some extraneous fields but the fields are ignored on the client side if the status is `noupdate`. The component updater code looks for the `update-check` status to display the result:
https://cs.chromium.org/chromium/src/components/update_client/component.cc?g=0&l=511

File: extension/extension.go (hunk 2 above)
Comment (jumde) [reply extension/extension.go]: When status is > 0, component number on the client has a version number higher than the component available on the server. I created a custom request for the google component updater to check the response and the server returns `noupdate`.

But I think returning error in this case is fine, since a lower component version on the client side is unexpected. This might be a result of malformed/malicious/corrupted files in the user directory.

Comment (bbondy) [reply extension/extension.go]: I think we should match it with noupdate, but we don't have to do that in the context of this task.
